lib/PlatoonService/src/V2VCommManager.cpp

```diff
@@ -0,0 +1,534 @@
+
+            if (0U == serializeJson(heartbeatDoc, heartbeatPayload))
+            {
+                LOG_ERROR("Failed to serialize heartbeat.");
+            }
+            else if (false == m_mqttClient.publish(m_heartbeatResponseTopic, false, heartbeatPayload))
+            {
+                LOG_ERROR("Failed to publish MQTT message to %s.", m_heartbeatResponseTopic.c_str());
+            }
+            else
+            {
+                LOG_DEBUG("Sent heartbeat: %s", heartbeatPayload.c_str());
+            }
+        }
+    }
+}
+
+void V2VCommManager::vehicleHeartbeatTopicCallback(const String& payload)
+{
+    /* Deserialize payload. */
+    StaticJsonDocument<JSON_HEARTBEAT_MAX_SIZE> jsonPayload;
+    DeserializationError                        error = deserializeJson(jsonPayload, payload.c_str());
+
+    if (DeserializationError::Ok != error)
+    {
+        LOG_ERROR("JSON Deserialization Error %d.", error);
+    }
+    else
+    {
+        JsonVariant jsonId        = jsonPayload["id"];        /* Vehicle ID. */
+        JsonVariant jsonTimestamp = jsonPayload["timestamp"]; /* Timestamp [ms]. */
+        JsonVariant jsonStatus    = jsonPayload["status"];    /* Vehicle status. */
+
+        if ((false == jsonId.isNull()) && (false == jsonTimestamp.isNull()) && (false == jsonStatus.isNull()))
+        {
+            uint8_t  id        = jsonId.as<uint8_t>();
+            uint32_t timestamp = jsonTimestamp.as<uint32_t>();
+            uint8_t  status    = jsonStatus.as<uint8_t>();
+
+            if (PLATOON_LEADER_ID == id)
+            {
+                /* This is me, the leader! */
+            }
+            else if (id > NUMBER_OF_FOLLOWERS)
+            {
+                LOG_ERROR("Received heartbeat from unknown vehicle %d.", id);
+            }
+            else if (timestamp != m_lastPlatoonHeartbeatTimestamp)
+            {
+                LOG_ERROR("Received heartbeat from vehicle %d with timestamp %d, expected %d.", id, timestamp,
+                          m_lastPlatoonHeartbeatTimestamp);
+            }
+            else
+            {
+                uint8_t   followerArrayIndex = id - 1U; /*Already checked id != 0U */
+                Follower& follower           = m_followers[followerArrayIndex];
+
+                /* Update follower. */
+                follower.setLastHeartbeatTimestamp(timestamp);
+                follower.setStatus(status);
+
+                ++m_followerResponseCounter;
+
+                /* Check follower status. */
+                if (VEHICLE_STATUS_OK != status)
+                {
+                    LOG_ERROR("Follower %d status: %d.", id, status);
+                    m_v2vStatus = V2V_STATUS_FOLLOWER_ERROR;
+                }
+            }
+        }
+    }
+}
+
+bool V2VCommManager::setupWaypointTopics(uint8_t platoonId, uint8_t vehicleId)
+{
+    bool    isSuccessful  = false;
+    uint8_t nextVehicleId = vehicleId + 1U; /* Output is published to next vehicle. */

```

Comment:
Is there always a next vehicle?

Reply:
In the case of the last follower, it is thought of a "virtual" next vehicle. This allows for feedback to the leader, when the leader subscribes to this topic. And the platoon can be expanded without any change in the code. In this way the exception handling is only for the leader and this is done separately, otherwise the last follower would also have to get its own exception handling

Reply:
Will be fixed in the next PR. Right now it works because the leade subscribes to the next vehicle after the last follower, but this will be changed once events are introduced

Reply:
ok